### PR TITLE
feat: start dotnet detection

### LINKED ISSUE #942: dotnet: extract format feature

### LINKED ISSUE #943: dotnet: extract arch feature

### LINKED ISSUE #944: dotnet: extract os feature (or not)
TODO: determine if there are dotnet headers that specify an underlying OS. seems like there must be something to enable p/invoke to be limited to Windows

### PR DESCRIPTION
**Pull request by mr-tz**
detect dotnet files and implement basic dnfile feature extractor

closes #942
closes #943 
closes #944

also refactoring around format and analysis branches

### Checklist

- [ ] No CHANGELOG update needed

- [ ] No new tests needed

- [ ] No documentation update needed

### CODE REVIEW

File: capa/helpers.py [hunk 1]
```diff
@@ -35,3 +37,9 @@ def is_runtime_ida():
 
 def assert_never(value: NoReturn) -> NoReturn:
     assert False, f"Unhandled value: {value} ({type(value).__name__})"
+
+
+def is_dotnet_file(pe: PE) -> bool:
```
Comment (mr-tz) [inline capa/helpers.py:42]: using pefile here, could use dnfile?

File: capa/features/extractors/dotnetfile.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import dnfile as dnfile
+
+from capa.features.common import ARCH_I386, ARCH_AMD64, Arch
+from capa.features.extractors.base_extractor import FeatureExtractor
+
+logger = logging.getLogger(__name__)
+
+
+def extract_file_os(**kwargs):
+    # TODO yield OS(OS_ANY), 0x0?
```
Comment (williballenthin) [inline capa/features/extractors/dotnetfile.py:12]: https://github.com/mandiant/capa/issues/944#issuecomment-1087994840

File: capa/features/extractors/dotnetfile.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    yield
+
+
+def extract_file_arch(pe, **kwargs):
+    # TODO also see 32 - Assembly Table?
+    if pe.net.Flags.CLR_32BITREQUIRED:
+        yield Arch(ARCH_I386), 0x0
+    else:
+        yield Arch(ARCH_AMD64), 0x0
```
Comment (williballenthin) [inline capa/features/extractors/dotnetfile.py:21]: https://github.com/mandiant/capa/issues/943#issuecomment-1087995352

File: capa/features/extractors/dotnetfile.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        raise NotImplementedError("PefileFeatureExtract can only be used to extract file features")
+
+    def extract_insn_features(self, f, bb, insn):
+        raise NotImplementedError("PefileFeatureExtract can only be used to extract file features")
+
+    def is_library_function(self, va):
+        raise NotImplementedError("PefileFeatureExtract can only be used to extract file features")
+
+    def get_function_name(self, va):
+        raise NotImplementedError("PefileFeatureExtract can only be used to extract file features")
```
Comment (williballenthin) [inline capa/features/extractors/dotnetfile.py:91]:
```suggestion
    def get_functions(self):
        raise NotImplementedError("DotnetfileFeatureExtract can only be used to extract file features")

    def extract_function_features(self, f):
        raise NotImplementedError("DotnetfileFeatureExtract can only be used to extract file features")

    def get_basic_blocks(self, f):
        raise NotImplementedError("DotnetfileFeatureExtract can only be used to extract file features")

    def extract_basic_block_features(self, f, bb):
        raise NotImplementedError("DotnetfileFeatureExtract can only be used to extract file features")

    def get_instructions(self, f, bb):
        raise NotImplementedError("DotnetfileFeatureExtract can only be used to extract file features")

    def extract_insn_features(self, f, bb, insn):
        raise NotImplementedError("DotnetfileFeatureExtract can only be used to extract file features")

    def is_library_function(self, va):
        raise NotImplementedError("DotnetfileFeatureExtract can only be used to extract file features")

    def get_function_name(self, va):
        raise NotImplementedError("DotnetfileFeatureExtract can only be used to extract file features")
```

File: capa/features/extractors/dotnetfile.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            yield feature, va
+
+
+GLOBAL_HANDLERS = (
+    # TODO extract_file_os,
+    extract_file_arch,
+)
+
+
+class DotnetfileFeatureExtractor(FeatureExtractor):
```
Comment (williballenthin) [inline capa/features/extractors/dotnetfile.py:46]:
```suggestion
class DnfileFeatureExtractor(FeatureExtractor):
```

??? up for discussion

File: capa/main.py [hunk 6]
```diff
@@ -933,6 +935,11 @@ def main(argv=None):
             logger.error("Input file '%s' is not a valid PE file: %s", args.sample, str(e))
             return E_CORRUPT_FILE
 
+        is_dotnet = is_dotnet_file(file_extractor.pe)
+        if is_dotnet:
+            # TODO try/except
+            file_extractor = capa.features.extractors.dotnetfile.DotnetfileFeatureExtractor(args.sample)
```
Comment (williballenthin) [inline capa/main.py:941]: #938: eventually probably want to collect a list of extractors to run so we can handle both native + .NET

i understand this logic will undergo further revisions, just hyperlinking issues

File: capa/main.py [hunk 7]
```diff
@@ -933,6 +935,11 @@ def main(argv=None):
             logger.error("Input file '%s' is not a valid PE file: %s", args.sample, str(e))
             return E_CORRUPT_FILE
 
+        is_dotnet = is_dotnet_file(file_extractor.pe)
```
Comment (mr-tz) [inline capa/main.py:938]: main is quite long an complex already, dotnet would add more branches in various places

File: capa/main.py [hunk 8]
```diff
@@ -950,21 +957,24 @@ def main(argv=None):
             logger.error("Input file '%s' is not a valid ELF file: %s", args.sample, str(e))
             return E_CORRUPT_FILE
 
+        if is_dotnet:
+            # TODO has_dotnetfile_limitation...
```
Comment (mr-tz) [inline capa/main.py:961]: do we really want to split dotnet and other file limitations?

File: capa/main.py [hunk 9]
```diff
@@ -950,21 +957,24 @@ def main(argv=None):
...
+        elif has_file_limitation(rules, pure_file_capabilities):
             # bail if capa encountered file limitation e.g. a packed binary
             # do show the output in verbose mode, though.
             if not (args.verbose or args.vverbose or args.json):
                 logger.debug("file limitation short circuit, won't analyze fully.")
                 return E_FILE_LIMITATION
 
     try:
-        if args.format == "pe" or (args.format == "auto" and taste.startswith(b"MZ")):
+        if not is_dotnet and (args.format == "pe" or (args.format == "auto" and taste.startswith(b"MZ"))):
```
Comment (mr-tz) [inline capa/main.py:973]: smells like this needs refactoring

File: capa/main.py (hunk 7 above)
Comment (mike-hunhoff) [reply capa/main.py:938]: hmm good question. dnfile invokes pefile so currently we are invoking pefile's parsing twice (when processing .NET PEs). I'm unsure if this is a performance hit great enough to warrant complicating main with more logic. I'd say for now we don't invoke dnfile until we know we are processing a .NET PE. So IMO using pefile here is fine.

File: capa/helpers.py (hunk 1 above)
Comment (mike-hunhoff) [reply capa/helpers.py:42]: hmm good question. dnfile invokes pefile so currently we are invoking pefile's parsing twice (when processing .NET PEs). I'm unsure if this is a performance hit great enough to warrant complicating main with more logic. I'd say for now we don't invoke dnfile until we know we are processing a .NET PE. So IMO using pefile here is fine.

File: capa/main.py [hunk 10]
```diff
@@ -973,6 +983,9 @@ def main(argv=None):
         format = "freeze"
         with open(args.sample, "rb") as f:
             extractor = capa.features.freeze.load(f.read())
+    elif is_dotnet:
+        # TODO extractor = capa.features.extractors.dotnet.extractor.DnFeatureExtractor(...)
```
Comment (mike-hunhoff) [inline capa/main.py:987]: do we want to move this to https://github.com/mandiant/capa/blob/57fe1e27b60fdd0993f2395c4a52a1dabd6e4794/capa/main.py#L426??

Comment (mr-tz) [reply capa/main.py]: yes, good idea, should we add a new `format` (just used internally or providable by the user also?)?

File: capa/features/extractors/dnfile_.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+    def is_dotnet_file(self) -> bool:
+        return bool(self.pe.net)
+
+    def get_base_address(self) -> int:
+        return self.pe.net.struct.EntryPointTokenOrRva
+
+    def extract_global_features(self):
+        with open(self.path, "rb") as f:
+            buf = f.read()
```
Comment (mike-hunhoff) [inline capa/features/extractors/dnfile_.py:93]: do we need a file read here? I think `dnfile.dnPE` has everything we need.

File: capa/features/extractors/dnfile_.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+        yield from extract_global_features(self.pe, buf)
+
+    def extract_file_features(self):
+        with open(self.path, "rb") as f:
+            buf = f.read()
+
+        yield from extract_file_features(self.pe, buf)
```
Comment (mike-hunhoff) [inline capa/features/extractors/dnfile_.py:101]: see comment above.

File: capa/features/extractors/dnfile_.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+)
+
+
+def extract_global_features(pe, buf):
+    """
+    extract global features from given workspace
+
+    args:
+      pe (pefile.PE): the parsed PE
+      buf: the raw sample bytes
```
Comment (mike-hunhoff) [inline capa/features/extractors/dnfile_.py:63]: need to update comment; see below for question on needing the raw bytes.

File: capa/features/extractors/dnfile_.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+        yield Arch(ARCH_ANY), 0x0
+
+
+def extract_file_features(pe, buf):
+    """
+    extract file features from given workspace
+
+    args:
+      pe (pefile.PE): the parsed PE
+      buf: the raw sample bytes
```
Comment (mike-hunhoff) [inline capa/features/extractors/dnfile_.py:36]: need to update comment; see below for question on needing raw bytes.

File: capa/helpers.py [hunk 15]
```diff
@@ -35,3 +45,71 @@ def is_runtime_ida():
...
+
+def log_unsupported_runtime_error():
+    logger.error("-" * 80)
+    logger.error(" Unsupported runtime or Python interpreter.")
+    logger.error(" ")
+    logger.error(" capa supports running under Python 2.7 using Vivisect for binary analysis.")
+    logger.error(" It can also run within IDA Pro, using either Python 2.7 or 3.5+.")
+    logger.error(" ")
+    logger.error(" If you're seeing this message on the command line, please ensure you're running Python 2.7.")
+    logger.error("-" * 80)
```
Comment (mike-hunhoff) [inline capa/helpers.py:115]: need to update these messages; it's been a long time since Python 2.7 😄

**Review by mike-hunhoff — CHANGES_REQUESTED** [review]
looks great! just a few questions to address.

File: capa/helpers.py (hunk 15 above)
Comment (mr-tz) [reply capa/helpers.py:115]: true!

File: capa/features/extractors/dnfile_.py (hunk 11 above)
Comment (mr-tz) [reply capa/features/extractors/dnfile_.py:93]: yes, as of now they're not needed and I will remove

Comment (mr-tz) [reply capa/features/extractors/dnfile_.py]: 239b03c8

File: capa/helpers.py (hunk 15 above)
Comment (mr-tz) [reply capa/helpers.py:115]: c91ac704

**Pull request merged**